File: 이코테/c++/전보.cpp
#include<bits/stdc++.h>
#define INF 1e9
using namespace std;

int N, M, C;
vector<pair<int, int> > graph[30001];
int dp[30001];

void dijkstra(int start){
	priority_queue<pair<int, int> > pq;
	pq.push({0, start});
	dp[start] = 0;
	while(!pq.empty()){
		int dist = -pq.top().first;
		int now = pq.top().second;
		pq.pop();
		if(dp[now] < dist) continue;
		for(int i=0; i<graph[now].size(); i++){
			int cost = graph[now][i].second + dist;
			if(cost < dp[graph[now][i].first]){
				dp[graph[now][i].first] = cost;
				pq.push(make_pair(-cost, graph[now][i].first));
			}
		}
	}
}

int main(void){
	cin >> N >> M >> C;
	for(int i=0; i<M; i++){
		int a, b, c;
		cin >> a >> b >> c;
		graph[a].push_back({b, c});
	}
	fill(dp, dp + 30001, INF);
	dijkstra(C);
	int count = 0;
	int max_dp = 0;
	for(int i=1; i<=N; i++){
		if(dp[i] != INF){
			count++;
			max_dp = max(max_dp, dp[i]);
		}
	}
	cout << count - 1 <<' ' << max_dp << endl;
}

File: 이코테 복습(python, c++, java)/최단 경로/python/전보.py
import heapq
import sys

input = sys.stdin.readline
INF = int(1e9)


def dijkstra(start):
    q = []
    heapq.heappush(q, (0, start))
    distance[start] = 0
    while q:
        dist, now = heapq.heappop(q)
        if distance[now] < dist:
            continue
        for i in graph[now]:
            cost = dist + i[1]
            if cost < distance[i[0]]:
                distance[i[0]] = cost
                heapq.heappush(q, (cost, i[0]))


N, M, C = map(int, input().split())
graph = [[] for _ in range(N + 1)]
distance = [INF] * (N + 1)
for _ in range(M):
    X, Y, Z = map(int, input().split())
    graph[X].append((Y, Z))

dijkstra(C)

count = 0
max_distance = 0
for d in distance:
    if d != 1e9:
        count += 1
        max_distance = max(max_distance, d)
print(count - 1, max_distance)
